simplefs/simplefs.go

```diff
@@ -1388,6 +1389,231 @@ func (k *SimpleFS) SimpleFSStat(ctx context.Context, path keybase1.Path) (de key
+	fi, err := fs.Lstat(finalElem)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	fipr, ok := fi.Sys().(libfs.PrevRevisionsGetter)
+	if !ok {
+		return nil, nil, simpleFSError{"Cannot get revisions for non-KBFS path"}
+	}
+	return fi, fipr.PrevRevisions(), nil
+}
+
+func (k *SimpleFS) doGetRevisions(
+	ctx context.Context, opID keybase1.OpID, path keybase1.Path,
+	spanType keybase1.RevisionSpanType) (
+	revs []keybase1.DirentWithRevision, err error) {
+	k.log.CDebugf(ctx, "Getting revisions for path %s, spanType=%s",
+		path, spanType)
+
+	// Both span types return 5 revisions.
+	k.setProgressTotals(opID, 0, 5)
+
+	fi, prs, err := k.getRevisionsFromPath(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+	if len(prs) == 0 {
+		return nil, simpleFSError{"No previous revisions"}
+	}
+
+	var currRev keybase1.DirentWithRevision
+	err = setStat(&currRev.Entry, fi)
+	if err != nil {
+		return nil, err
+	}
+	currRev.Revision = keybase1.KBFSRevision(prs[0].Revision)
+	k.log.CDebugf(ctx, "Found current revision: %d", prs[0].Revision)
+	k.updateReadProgress(opID, 0, 1)
+
+	var revPaths []keybase1.Path
+
+	// The next four depend on the span type.
+	pathStr := path.String()
+	switch spanType {
+	case keybase1.RevisionSpanType_DEFAULT:
+		// Use `prs` for the rest of the paths.
+		for i := 1; i < len(prs); i++ {
+			p := keybase1.NewPathWithKbfsArchived(keybase1.KBFSArchivedPath{
+				Path: pathStr,
+				ArchivedParam: keybase1.NewKBFSArchivedParamWithRevision(
+					keybase1.KBFSRevision(prs[i].Revision)),
+			})
+			revPaths = append(revPaths, p)
+		}
+	case keybase1.RevisionSpanType_LAST_FIVE:
+		expectedCount := uint8(2)
+		nextSlot := 1
+		lastRevision := prs[0].Revision
+
+		// Step back through the previous revisions.  If the next one
+		// in the list happens to be the next in line (because the
+		// count is one more than the current count), use it.
+		// Otherwise, we have to fetch the stats from the MD revision
+		// before the last one we processed, and use the
+		// PreviousRevisions list from that version of the file.
+		for len(revPaths) < 4 && nextSlot < len(prs) {
+			var rev kbfsmd.Revision
+			if prs[nextSlot].Count == expectedCount {
+				rev = prs[nextSlot].Revision
+			} else if lastRevision > kbfsmd.RevisionInitial {
+				k.log.CDebugf(ctx, "Inspecting revision %d to find previous",
+					lastRevision-1)
+				pathToPrev := keybase1.NewPathWithKbfsArchived(
+					keybase1.KBFSArchivedPath{
+						Path: pathStr,
+						ArchivedParam: keybase1.NewKBFSArchivedParamWithRevision(
+							keybase1.KBFSRevision(lastRevision - 1)),
+					})
+				_, prevPRs, err := k.getRevisionsFromPath(ctx, pathToPrev)
+				if _, isGC := err.(libkbfs.RevGarbageCollectedError); isGC {
+					k.log.CDebugf(ctx, "Hit a GC'd revision: %d",
+						lastRevision-1)
+					break
+				} else if err != nil {
+					return nil, err
+				}
+				if len(prevPRs) == 0 {
+					return nil, simpleFSError{"No previous revisions"}
```

Comment:
Why is this an error? Shouldn't we still return if there are only 3 revisions?

Reply:
It's because if the file can be stat'd, the previous revisions should always include at least it's own revision (since we're looking at a revision from a previous list, so we know the file should be there). So I guess this error text is a bit confusing, I'll change the text.

Reply:
Let me amend that slightly: we looking up `lastRevision - 1` here, but we know there's another slot in the previous revisions list at this point (because `nextSlot < len(prs)`, so we know that there is _some_ other revision between the last one and the next one, so there should definitely be a stat'able file and a non-empty previous revisions list here.

Reply:
Got it, that makes much more sense than my interpretation. Maybe a comment explaining that here would help.

Reply:
Sure, done in https://github.com/keybase/kbfs/commit/e0c60848371b4e0169b2cd81b8aff79c06bdc68b.